torch_xla/csrc/dynamic_shape_detector.cpp

```diff
@@ -0,0 +1,254 @@
+
+    // Create the children_ map entry for the newly created suffix node.
+    children_[suffix.front().hash] = std::move(suffix_node);
+
+    TF_VLOG(5) << "Split node " << children_[suffix.front().hash].get()
+               << " at position " << matched << ": " << suffix.front().str
+               << " (" << suffix.front().hash << ")";
+  }
+
+  // This node's common_sequence_ will be whatever the prefix was.
+  common_sequence_ = std::vector<TrieValue>{prefix.begin(), prefix.end()};
+}
+
+DynamicShapeDetector* DynamicShapeDetector::Get() {
+  static DynamicShapeDetector ds_detector = DynamicShapeDetector();
+  return &ds_detector;
+}
+
+void DynamicShapeDetector::StartSession(std::string session) {
+  if (session_infos_.find(session) == session_infos_.end()) {
+    // Create a new session, with a fresh TrieNode.
+    session_infos_[session] = {session, std::make_unique<TrieNode>(), 0};
+    TF_VLOG(5) << "Created new session: " << session;
+  }
+  current_session_ = &session_infos_[session];
+  TF_VLOG(5) << "Started session: " << session;
+  RootBuilder();
+}
+
+void DynamicShapeDetector::SetMaxAllowedTraces(std::size_t value) {
+  max_allowed_traces_per_function = value;
+}
+
+std::size_t DynamicShapeDetector::GetMaxAllowedTraces() {
+  return max_allowed_traces_per_function;
+}
+
+bool DynamicShapeDetector::IsSessionActive() {
+  return current_session_ != nullptr;
+}
+
+bool DynamicShapeDetector::AllowNewTrace() {
+  XLA_CHECK(IsSessionActive());
+  return current_session_->traces_ < max_allowed_traces_per_function;
+}
+
+void DynamicShapeDetector::EndSession() {
+  XLA_CHECK(IsSessionActive());
+
+  try {
+    // Mark the current builder_ node as trace boundary.
+    // If we did create a new trace, increment the session's trace number.
+    if (builder_.MarkTraceBoundary(AllowNewTrace())) {
+      current_session_->traces_++;
+      TF_VLOG(5) << "Created new trace.";
+    }
+
+    ResetSession();
+    TF_VLOG(5) << "Ended session: " << current_session_->name_;
+  } catch (const std::exception& e) {
+    // MarkTraceBoundary might raise an exception if AllowNewTrace() is false.
+    // Catch it here, so that we can correctly end the session.
+    ResetSession();
+    throw;
+  }
+}
+
+void DynamicShapeDetector::ResetSession() {
+  current_session_ = nullptr;
+  builder_ = {};
+}
+
+void DynamicShapeDetector::RootBuilder() {
+  builder_ = current_session_->NewBuilder();
+}
+
+void DynamicShapeDetector::AddNodeInfo(torch::lazy::hash_t hash,
+                                       const std::string& str) {
+  XLA_CHECK(current_session_ != nullptr);
+
+  try {
+    builder_.AddValue({hash, str}, AllowNewTrace());
+  } catch (const std::exception& e) {
+    // AddValue might raise an exception if AllowNewTrace() is false. Catch it
+    // here, so that we can correctly return the builder to the root of the
+    // trie.
+    //
+    // TODO(ysiraichi): we should actually rollback this trace.
```

Comment:
As is, it should work as we expect. However, a better approach (for a future PR) would be to rollback the changes introduced since the last session start.

Reply:
if we hit a error likely the program will just end so maybe it is fine.